harness/api_checkers.go

```diff
@@ -0,0 +1,437 @@
+func (s *APICheckSuite) TestSpanBaggage() {
+	span := s.tracer.StartSpan("Fry")
+	assertEmptyBaggage(s.T(), span.Context())
+
+	spanRef := span.SetBaggageItem("Kiff-loves", "Amy")
+	assert.Exactly(s.T(), spanRef, span)
+
+	val := span.BaggageItem("Kiff-loves")
+	if s.opts.CheckBaggageValues {
+		assert.Equal(s.T(), "Amy", val)
+	} else {
+		s.T().Log("Baggage propagation not supported, not checking")
+	}
+	span.Finish()
+}
+
+// TestContextBaggage tests calls to set and get span baggage, and if the CheckBaggageValues option
+// is set, asserts that baggage values were successfully retrieved from the span's SpanContext.
+func (s *APICheckSuite) TestContextBaggage() {
+	span := s.tracer.StartSpan("Fry")
+	assertEmptyBaggage(s.T(), span.Context())
+
+	span.SetBaggageItem("Kiff-loves", "Amy")
+	if s.opts.CheckBaggageValues {
+		called := false
+		span.Context().ForeachBaggageItem(func(k, v string) bool {
+			assert.False(s.T(), called)
+			called = true
+			assert.Equal(s.T(), "Kiff-loves", k)
+			assert.Equal(s.T(), "Amy", v)
+			return true
+		})
+	} else {
+		s.T().Log("Baggage propagation not supported, not checking")
+	}
+	span.Finish()
+}
+
+// TestTextPropagation tests if the Tracer can Inject a span into a TextMapCarrier, and later Extract it.
+// If CheckExtract is set, it will check if Extract was successful (returned no error). If a Probe is set,
+// it will check if the extracted context is in the same trace as the original span.
+func (s *APICheckSuite) TestTextPropagation() {
+	span := s.tracer.StartSpan("Bender")
+	textCarrier := opentracing.TextMapCarrier{}
+	err := span.Tracer().Inject(span.Context(), opentracing.TextMap, textCarrier)
+	assert.NoError(s.T(), err)
+
+	extractedContext, err := s.tracer.Extract(opentracing.TextMap, textCarrier)
+	if s.opts.CheckExtract {
+		assert.NoError(s.T(), err)
+		assertEmptyBaggage(s.T(), extractedContext)
+	} else {
+		s.T().Log("Tracer.Extract not supported, not checking")
+	}
+	if s.opts.Probe != nil {
+		assert.True(s.T(), s.opts.Probe.SameSpanContext(span, extractedContext))
+	}
+	span.Finish()
+}
+
+// TestHTTPPropagation tests if the Tracer can Inject a span into HTTP headers, and later Extract it.
+// If CheckExtract is set, it will check if Extract was successful (returned no error). If a Probe is set,
+// it will check if the extracted context is in the same trace as the original span.
+func (s *APICheckSuite) TestHTTPPropagation() {
+	span := s.tracer.StartSpan("Bender")
+	textCarrier := opentracing.HTTPHeadersCarrier{}
+	// XXX add same test cases around valid HTTP header characters, casing
+	err := span.Tracer().Inject(span.Context(), opentracing.HTTPHeaders, textCarrier)
+	assert.NoError(s.T(), err)
+
+	extractedContext, err := s.tracer.Extract(opentracing.HTTPHeaders, textCarrier)
+	if s.opts.CheckExtract {
+		assert.NoError(s.T(), err)
+		assertEmptyBaggage(s.T(), extractedContext)
+	} else {
+		s.T().Log("Tracer.Extract not supported, skipping")
+	}
+	if s.opts.Probe != nil {
```

Comment:
Hrmmm I can definitely add else Log calls everywhere I missed ... but to do what you're describing I think it would be much easier to use Go 1.7 subtests to call `t.Skip` from a subtest rather than refactoring / ripping apart the test cases into separate test functions depending on capabilities.